feat: add `iter/cartesian-power` (#1350)
stdlib-js/stdlib: node_modules/@stdlib/iter/cartesian-power/README.md

<!--

@license Apache-2.0

Copyright (c) 2024 The Stdlib Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

-->

# iterCartesianPower

> > Create an iterator which generates the Cartesian power of an input array-like object.

Reviewer: ```suggestion
> Create an iterator which returns the Cartesian power.
```

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- Package usage documentation. -->

<section class="usage">

## Usage

```javascript
var iterCartesianPower = require( '@stdlib/iter/cartesian-power' );
```

#### iterCartesianPower( x, n )

> Returns an iterator which generates the Cartesian power of an input array-like object.

Reviewer: ```suggestion
Returns an iterator which returns the Cartesian power.
```

```javascript

> var x = ['a', 'b', 'c'];
> var n = 2;
> var iterator = iterCartesianPower( x, n );

Reviewer: ```suggestion
var x = [ 'a', 'b', 'c' ];

var iterator = iterCartesianPower( x, 2 );
```

> for (const combination of iterator) {

Reviewer: Everything should be written in ES5. See also https://github.com/stdlib-js/stdlib/blob/develop/FAQ.md#es2015-and-beyond.

Reviewer: You can perform manual iteration. See other iterator packages for examples.

    console.log( combination );
}

```

</section>

<!-- /.usage -->

<!-- Package usage notes. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="notes">

> ## Notes
> 
> - The function expects the first argument x to be an array-like object.
> - The second argument n must be a non-negative integer.
> - The returned iterator will generate all possible combinations of n elements from the input array x.
>     

Reviewer: ```suggestion
```

We don't typically explain edge cases. For the last note, this is implied by the concept of the Cartesian power.

</section>

<!-- /.notes -->

<!-- Package usage examples. -->

<section class="examples">

## Examples

> 

Reviewer: ```suggestion
```

Do not include spurious newlines unless specified to do so by linting.

<!-- eslint no-undef: "error" -->

```javascript
// Example 1: Generating Cartesian power of an array with n = 2
var x = [1, 2, 3];
var n = 2;
var iterator = iterCartesianPower(x, n);

> for (const combination of iterator) {

Reviewer: ES5 here and elsewhere.

    console.log(combination);
}

// Output:
// [ 1, 1 ]
// [ 1, 2 ]
// [ 1, 3 ]
// [ 2, 1 ]
// [ 2, 2 ]
// [ 2, 3 ]
// [ 3, 1 ]
// [ 3, 2 ]
// [ 3, 3 ]

```

> ```javascript

Reviewer: Don't split up examples like this. This does not match the conventions in this project.

var x = ['a', 'b'];
var n = 3;
var iterator = iterCartesianPower(x, n);

for (const combination of iterator) {
    console.log(combination);
}

// Output:
// [ 'a', 'a', 'a' ]
// [ 'a', 'a', 'b' ]
// [ 'a', 'b', 'a' ]
// [ 'a', 'b', 'b' ]
// [ 'b', 'a', 'a' ]
// [ 'b', 'a', 'b' ]
// [ 'b', 'b', 'a' ]
// [ 'b', 'b', 'b' ]

```

```javascript
var x = ['a', 'b', 'c'];
var n = 1;
var iterator = iterCartesianPower(x, n);

for (const combination of iterator) {
    console.log(combination);
}

// Output:
// [ 'a' ]
// [ 'b' ]
// [ 'c' ]
```

```javascript
var x = ['a', 'b', 'c'];
var n = 0;
var iterator = iterCartesianPower(x, n);

for (const combination of iterator) {
    console.log(combination);
}

// Output:
// []
```

</section>

<!-- /.examples -->

<!-- Section to include cited references. If references are included, add a horizontal rule *before* the section. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="references">

</section>

<!-- /.references -->

<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->

> <section class="related">
> 
> * * *
> 
> ## See Also
> 
> -   <span class="package-name">[`@stdlib/array/from-iterator`][@stdlib/array/from-iterator]</span><span class="delimiter">: </span><span class="description">create (or fill) an array from an iterator.</span>
> -   <span class="package-name">[`@stdlib/iter/datespace`][@stdlib/iter/datespace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns evenly spaced dates over a specified interval.</span>
> -   <span class="package-name">[`@stdlib/iter/incrspace`][@stdlib/iter/incrspace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns evenly spaced numbers according to a specified increment.</span>
> -   <span class="package-name">[`@stdlib/iter/logspace`][@stdlib/iter/logspace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns evenly spaced numbers on a log scale.</span>
> -   <span class="package-name">[`@stdlib/iter/step`][@stdlib/iter/step]</span><span class="delimiter">: </span><span class="description">create an iterator which returns a sequence of numbers according to a specified increment.</span>
> -   <span class="package-name">[`@stdlib/iter/unitspace`][@stdlib/iter/unitspace]</span><span class="delimiter">: </span><span class="description">create an iterator which returns numbers incremented by one.</span>

Reviewer: ```suggestion
<section class="related">
```

</section>

<!-- /.related -->

<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->

<section class="links">

[@stdlib/math/base/special/roundn]: https://www.npmjs.com/package/@stdlib/math-base-special-roundn

> <!-- <related-links> -->
> 
> [@stdlib/array/from-iterator]: https://www.npmjs.com/package/@stdlib/array-from-iterator
> 
> [@stdlib/iter/datespace]: https://github.com/stdlib-js/iter/tree/main/datespace
> 
> [@stdlib/iter/incrspace]: https://github.com/stdlib-js/iter/tree/main/incrspace
> 
> [@stdlib/iter/logspace]: https://github.com/stdlib-js/iter/tree/main/logspace
> 
> [@stdlib/iter/step]: https://github.com/stdlib-js/iter/tree/main/step
> 
> [@stdlib/iter/unitspace]: https://github.com/stdlib-js/iter/tree/main/unitspace
> 
> <!-- </related-links> -->
> 

Reviewer: ```suggestion
```
